File: libs/sdk-bindings/bindings-android/lib/src/main/kotlin/breez_sdk_notification/job/ConfirmTransaction.kt
package breez_sdk_notification.job

import android.content.Context
import breez_sdk.BlockingBreezServices
import breez_sdk.BreezEvent
import breez_sdk.ReverseSwapStatus
import breez_sdk_notification.Constants.DEFAULT_SWAP_TX_CONFIRMED_NOTIFICATION_FAILURE_TEXT
import breez_sdk_notification.Constants.DEFAULT_SWAP_TX_CONFIRMED_NOTIFICATION_FAILURE_TITLE
import breez_sdk_notification.Constants.DEFAULT_SWAP_TX_CONFIRMED_NOTIFICATION_TITLE
import breez_sdk_notification.Constants.NOTIFICATION_CHANNEL_ADDRESS_TXS_CONFIRMED
import breez_sdk_notification.Constants.SWAP_TX_CONFIRMED_NOTIFICATION_FAILURE_TEXT
import breez_sdk_notification.Constants.SWAP_TX_CONFIRMED_NOTIFICATION_FAILURE_TITLE
import breez_sdk_notification.Constants.SWAP_TX_CONFIRMED_NOTIFICATION_TITLE
import breez_sdk_notification.NotificationHelper.Companion.notifyChannel
import breez_sdk_notification.ResourceHelper.Companion.getString
import breez_sdk_notification.SdkForegroundService
import breez_sdk_notification.ServiceLogger
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
data class AddressTxsConfirmedRequest(
    val address: String,
)

class ConfirmTransactionJob(
    private val context: Context,
    private val fgService: SdkForegroundService,
    private val payload: String,
    private val logger: ServiceLogger,
    private var bitcoinAddress: String? = null,
) : Job {
    companion object {
        private const val TAG = "ConfirmTransactionJob"
    }

    override fun start(breezSDK: BlockingBreezServices) {
        try {
            val request = Json.decodeFromString(AddressTxsConfirmedRequest.serializer(), payload)
            this.bitcoinAddress = request.address
        } catch (e: Exception) {
            logger.log(TAG, "Failed to decode payload: ${e.message}", "WARN")
        }

        this.bitcoinAddress?.let {address ->
            try {
                breezSDK.redeemSwap(address)
                logger.log(TAG, "Found swap for $address", "INFO")
                return
            } catch (e: Exception) {
                logger.log(TAG, "Failed to redeem swap: ${e.message}", "WARN")
            }

            try {
                breezSDK.claimReverseSwap(address)
                logger.log(TAG, "Found reverse swap for $address", "INFO")
                return
            } catch (e: Exception) {
                logger.log(TAG, "Failed to process reverse swap: ${e.message}", "WARN")
            }
        }

        fgService.onFinished(this)
    }

    override fun onEvent(e: BreezEvent) {
        this.bitcoinAddress?.let {address ->
            when (e) {
                is BreezEvent.ReverseSwapUpdated -> {
                    val revSwapInfo = e.details
                    logger.log(TAG, "Received reverse swap updated event: ${revSwapInfo.id} current address: $address status: ${revSwapInfo.status}", "TRACE")
                    if (revSwapInfo.status == ReverseSwapStatus.COMPLETED_SEEN || revSwapInfo.status == ReverseSwapStatus.COMPLETED_CONFIRMED) {
                        notifySuccess(address)
                    }
                }

                is BreezEvent.SwapUpdated -> {
                    val swapInfo = e.details
                    logger.log(TAG, "Received swap updated event: ${swapInfo.bitcoinAddress} current address: $address status: ${swapInfo.status}", "TRACE")
                    if (swapInfo.bitcoinAddress == address) {
                        if (swapInfo.paidMsat.toLong() > 0) {
                            notifySuccess(address)
                        }
                    }
                }

                else -> {}
            }
        }
    }

    override fun onShutdown() {
        notifyFailure()
    }

    private fun notifySuccess(address: String) {
        logger.log(TAG, "Address $address processed successfully", "INFO")
        notifyChannel(
            context,
            NOTIFICATION_CHANNEL_ADDRESS_TXS_CONFIRMED,
            getString(
                context,
                SWAP_TX_CONFIRMED_NOTIFICATION_TITLE,
                DEFAULT_SWAP_TX_CONFIRMED_NOTIFICATION_TITLE
            ),
        )
        fgService.onFinished(this)
    }

    private fun notifyFailure() {
        this.bitcoinAddress?.let{address -> 
            logger.log(TAG, "Address $address processing failed", "INFO")
            notifyChannel(
                context,
                NOTIFICATION_CHANNEL_ADDRESS_TXS_CONFIRMED,
                getString(
                    context,
                    SWAP_TX_CONFIRMED_NOTIFICATION_FAILURE_TITLE,
                    DEFAULT_SWAP_TX_CONFIRMED_NOTIFICATION_FAILURE_TITLE
                ),
                getString(
                    context,
                    SWAP_TX_CONFIRMED_NOTIFICATION_FAILURE_TEXT,
                    DEFAULT_SWAP_TX_CONFIRMED_NOTIFICATION_FAILURE_TEXT
                ),
            )
        }
    }
}

File: libs/sdk-bindings/bindings-swift/Sources/BreezSDK/Task/ConfirmTransaction.swift
import UserNotifications
import Foundation

struct AddressTxsConfirmedRequest: Codable {
    let address: String
}

class ConfirmTransactionTask : TaskProtocol {
    fileprivate let TAG = "ConfirmTransactionTask"
    
    internal var payload: String
    internal var contentHandler: ((UNNotificationContent) -> Void)?
    internal var bestAttemptContent: UNMutableNotificationContent?
    internal var logger: ServiceLogger
    internal var bitcoinAddress: String? = nil
    
    init(payload: String, logger: ServiceLogger, contentHandler: ((UNNotificationContent) -> Void)? = nil, bestAttemptContent: UNMutableNotificationContent? = nil) {
        self.payload = payload
        self.contentHandler = contentHandler
        self.bestAttemptContent = bestAttemptContent
        self.logger = logger
    }
    
    public func onEvent(e: BreezEvent) {
        if let address = self.bitcoinAddress {
            switch e {
            case .reverseSwapUpdated(details: let revSwapInfo):
                self.logger.log(tag: TAG, line: "Received reverse swap updated event: \(revSwapInfo.id), current address: \(address) status: \(revSwapInfo.status)\n", level: "INFO")
                if case .completedSeen = revSwapInfo.status, case  .completedConfirmed = revSwapInfo.status {
                    self.notifySuccess()
                }
                break
            case .swapUpdated(details: let swapInfo):
                self.logger.log(tag: TAG, line: "Received swap updated event: \(swapInfo.bitcoinAddress), current address: \(address) status: \(swapInfo.status)\n", level: "INFO")
                if address == swapInfo.bitcoinAddress {
                    if (swapInfo.paidMsat > 0) {
                        self.notifySuccess()
                    }
                }
                break
            default:
                break
            }
        }
    }
    
    func start(breezSDK: BlockingBreezServices) throws {
        do {
            let addressTxsConfirmedRequest = try JSONDecoder().decode(AddressTxsConfirmedRequest.self, from: self.payload.data(using: .utf8)!)
            self.bitcoinAddress = addressTxsConfirmedRequest.address
        } catch let e {
            self.logger.log(tag: TAG, line: "Failed to decode payload: \(e)", level: "ERROR")
            self.onShutdown()
            throw e
        }
        
        guard let address = bitcoinAddress else {
            self.logger.log(tag: TAG, line: "Address not in payload", level: "ERROR")
            self.onShutdown()
            return
        }
        
        do {
            try breezSDK.redeemSwap(swapAddress: address)
            self.logger.log(tag: TAG, line: "Found swap for \(address)", level: "DEBUG")
            return
        } catch let e {
            self.logger.log(tag: TAG, line: "Failed to redeem swap: \(e)", level: "ERROR")
        }
        
        do {
            try breezSDK.claimReverseSwap(lockupAddress: address)
            self.logger.log(tag: TAG, line: "Found reverse swap for \(address)", level: "DEBUG")
        } catch let e {
            self.logger.log(tag: TAG, line: "Failed to process reverse swap: \(e)", level: "ERROR")
        }
    }

    func onShutdown() {
        let notificationTitle = ResourceHelper.shared.getString(key: Constants.SWAP_TX_CONFIRMED_NOTIFICATION_FAILURE_TITLE, fallback: Constants.DEFAULT_SWAP_TX_CONFIRMED_NOTIFICATION_FAILURE_TITLE)
        let notificationBody = ResourceHelper.shared.getString(key: Constants.SWAP_TX_CONFIRMED_NOTIFICATION_FAILURE_TEXT, fallback: Constants.DEFAULT_SWAP_TX_CONFIRMED_NOTIFICATION_FAILURE_TEXT)
        self.displayPushNotification(title: notificationTitle, body: notificationBody, logger: self.logger, threadIdentifier: Constants.NOTIFICATION_THREAD_ADDRESS_TXS_CONFIRMED)
    }

    func notifySuccess() {
        self.logger.log(tag: TAG, line: "Address \(self.bitcoinAddress) processed successfully", level: "INFO")
        let notificationTitle = ResourceHelper.shared.getString(key: Constants.SWAP_TX_CONFIRMED_NOTIFICATION_TITLE, fallback: Constants.DEFAULT_SWAP_TX_CONFIRMED_NOTIFICATION_TITLE)
        self.displayPushNotification(title: notificationTitle, logger: self.logger, threadIdentifier: Constants.NOTIFICATION_THREAD_ADDRESS_TXS_CONFIRMED)
    }
}
